crates/scheduler/src/evaluators.rs

```diff
@@ -0,0 +1,372 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::sync::Arc;
+use std::time::Instant;
+
+use snafu::{OptionExt, ResultExt};
+use tokio::task::JoinHandle;
+use tokio_util::sync::CancellationToken;
+use uuid::Uuid;
+
+use crate::tasks::{TaskDelivery, TaskRequest};
+use crate::{Error, Result};
+
+#[derive(PartialEq)]
+pub enum EvaluatorType {
+    /// The evaluator returns the next request in the schedule, only if the current time is greater than the last request.
+    /// E.g. `.next()` -> 1am, at 12:30am `.next()` -> 1am, at 1:00 am `.next()` -> 2am.
+    Sequential,
+    /// The evaluator returns requests based on a fixed interval, anchored from a specific time.
+    /// E.g. `.next()` -> 1am, at 12:30am `.next()` -> 1:30am, at 1:00 am `.next()` -> 2:00am.
+    /// Resetting a timed evaluator should reset the anchor time to the current time.
+    Timed,
+    /// The evaluator returns interrupt requests.
+    /// Interrupts have the highest priority and should be delivered immediately.
+    /// The scheduler should abort any in-progress evaluation wait timers and deliver the interrupt request.
+    Interrupt,
+}
+
+pub trait Evaluator: Send + Sync {
+    /// Returns a unique identifier for this evaluator.
+    fn id(&self) -> Arc<Uuid>;
+    /// Returns the type of this evaluator.
+    fn evaluator_type(&self) -> EvaluatorType;
+    /// Returns the next task request for this evaluator.
+    fn evaluate(&mut self) -> Option<Arc<TaskRequest>>;
+    /// Clears the internal state of this evaluator to reset it to its initial state.
+    fn reset(&mut self);
+}
+
+pub struct IntervalEvaluator {
+    id: Arc<Uuid>,
+    interval: u64,
```

Comment:
Can you use `std::time::Duration` here instead? Then you do not have to convert to seconds `Duration::.as_secs()` below and elsewhere.